### PR TITLE
Enable TLS on lens-k8s-proxy

### PR DESCRIPTION
**Pull request by jakolehm**
Enables secure TLS connections from clients (mainly `LensProxy`) to `lens-k8s-proxy`. Follow-up PR will enable TLS also from clients (renderer etc.) to `LensProxy`.

### CODE REVIEW

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ */
+import { getInjectable } from "@ogre-tools/injectable";
+import directoryForUserDataInjectable from "../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
+import { createKubeAuthProxyCertificateFiles, getKubeAuthProxyCertificatePath } from "./kube-auth-proxy-cert";
+
+const getKubeAuthProxyCertDirInjectable = getInjectable({
+  id: "get-kube-auth-proxy-cert-dir",
+
+  setup: async (di) => {
+    await createKubeAuthProxyCertificateFiles(getKubeAuthProxyCertificatePath(await di.inject(directoryForUserDataInjectable)));
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts]: Hmm. Add variables? One line of code should do one thing. Also code should be left-to-right readable :)

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  },
+
+  instantiate: (di) => getKubeAuthProxyCertificatePath(di.inject(directoryForUserDataInjectable)),
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts]: Add variables.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+export async function createKubeAuthProxyCertificateFiles(dir: string): Promise<string> {
+  const cert = getKubeAuthProxyCertificate();
+
+  try {
+    await access(dir);
+  } catch {
+    await mkdir(dir);
+  }
+
+  await writeFile(path.join(dir, "proxy.key"), cert.private);
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Explicit side-effect `writeFile` should be a dependency instead.

File: src/main/kube-auth-proxy/kube-auth-proxy.ts [hunk 4]
```diff
@@ -42,13 +43,15 @@ export class KubeAuthProxy {
     }
 
     const proxyBin = this.dependencies.proxyBinPath;
-    
-    this.proxyProcess = spawn(proxyBin, [], { 
+    const certPath = await this.dependencies.proxyCertPath;
+
+    this.proxyProcess = spawn(proxyBin, [], {
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy.ts]: Explicit side-effect `spawn` should be a dependency instead.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+export function getKubeAuthProxyCertificatePath(baseDir: string) {
+  return path.join(baseDir, "kube-auth-proxy");
+}
+
+export async function createKubeAuthProxyCertificateFiles(dir: string): Promise<string> {
+  const cert = getKubeAuthProxyCertificate();
+
+  try {
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Avoidable try/catch as part of business-logic, which is a code-smell. Try/catch is best reserved for fatal errors, and using it for something else damages credibility of eg. centralized error handling.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    await access(dir);
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Explicit side-effect `access` should be a dependency instead.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  } catch {
+    await mkdir(dir);
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Explicit side-effect `mkdir` should be a dependency instead.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+export function getKubeAuthProxyCertificatePath(baseDir: string) {
+  return path.join(baseDir, "kube-auth-proxy");
+}
+
+export async function createKubeAuthProxyCertificateFiles(dir: string): Promise<string> {
+  const cert = getKubeAuthProxyCertificate();
+
+  try {
+    await access(dir);
+  } catch {
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: "catch all" could be "catch specific" if decision is to use `catch` here.

File: src/main/context-handler/context-handler.ts [hunk 9]
```diff
@@ -132,10 +135,21 @@ export class ContextHandler {
   }
 
   protected async newApiTarget(timeout: number): Promise<httpProxy.ServerOptions> {
+    await this.ensureServer();
```
Comment (Iku-turso) [inline src/main/context-handler/context-handler.ts:141]: Explicit side-effect `ensureServer` should be a dependency instead.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
+import { access, mkdir, writeFile } from "fs/promises";
+import path from "path";
+import * as selfsigned from "selfsigned";
+
+export function getKubeAuthProxyCertificate(): selfsigned.SelfSignedCert {
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Does not require to be a function?

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  const opts = [
+    { name: "commonName", value: "Lens Certificate Authority" },
+    { name: "organizationName", value: "Lens" },
+  ];
+
+  return selfsigned.generate(opts, {
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Is `selfSigned.generate` deterministic? If not, it is an explicit side-effect, and should be a dependency instead.

File: src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts [hunk 12]
```diff
@@ -17,6 +18,7 @@ const createKubeAuthProxyInjectable = getInjectable({
     const proxyPath = isDevelopment ? path.join("client", process.platform, process.arch) : process.arch;
     const dependencies = {
       proxyBinPath: path.join(di.inject(directoryForBundledBinariesInjectable), proxyPath, binaryName),
+      proxyCertPath: di.inject(getKubeAuthProxyCertDirInjectable),
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts]: Name `getKubeAuthProxyCertDirInjectable` implies function, but it is used as a value here.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      { name: "basicConstraints", cA: true },
+      { name: "subjectAltName", altNames: [
+        { type: 2, value: "localhost" },
+        { type: 7, ip: "127.0.0.1" },
+      ] },
+    ],
+  });
+}
+
+export function getKubeAuthProxyCertificatePath(baseDir: string) {
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Thought: a convention is to use `function` to advertise that `this` has special significance. Here it does not, and lambda could be used to advertise that instead.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+import { getInjectable } from "@ogre-tools/injectable";
+import directoryForUserDataInjectable from "../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
+import { createKubeAuthProxyCertificateFiles, getKubeAuthProxyCertificatePath } from "./kube-auth-proxy-cert";
+
+const getKubeAuthProxyCertDirInjectable = getInjectable({
```
Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts]: Name of `injectable` does not document unrelated stuff done in `setup`.

Comment (Iku-turso) [inline src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts]: This could be wonderfully unit-testable either locally, or preferably as part of a larger system ;)

File: src/main/context-handler/context-handler.ts [hunk 15]
```diff
@@ -27,6 +29,7 @@ interface PrometheusServicePreferences {
 
 interface Dependencies {
   createKubeAuthProxy: (cluster: Cluster, environmentVariables: NodeJS.ProcessEnv) => KubeAuthProxy;
+  certPath: string;
 }
 
 export class ContextHandler {
```
Comment (Iku-turso) [inline src/main/context-handler/context-handler.ts:34]: Note: some unrelated janky code around these parts needs a lovin' champion at some.

File: src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts (hunk 12 above)
Comment (jakolehm) [reply src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts]: What would be better naming convention for values? `createKubeAuthProxyCertDirInjectable`?

File: src/main/context-handler/context-handler.ts (hunk 9 above)
Comment (jakolehm) [reply src/main/context-handler/context-handler.ts:141]: Might be out-of-scope for this PR (not sure yet).

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts (hunk 14 above)
Comment (jakolehm) [reply src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts]: What would be better naming convention for injectables that have setup?

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts (hunk 1 above)
Comment (jakolehm) [reply src/main/kube-auth-proxy/kube-auth-proxy-cert.injectable.ts]: Yes, will cleanup this PR before removing "draft".

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts (hunk 11 above)
Comment (jakolehm) [reply src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: Not 100% deterministic because cert creation always creates an unique certificate. That said, it has deterministic fields... not sure where we should draw the line here.

File: src/main/kube-auth-proxy/kube-auth-proxy-cert.ts (hunk 13 above)
Comment (Nokel81) [reply src/main/kube-auth-proxy/kube-auth-proxy-cert.ts]: I would disagree, this codebase only uses `class`-es for `this` significant code and lambdas can sometimes make the code harder to read.

File: src/main/index.ts [hunk 16]
```diff
@@ -62,9 +62,11 @@ const di = getDi();
 
 app.setName(appName);
 
-di.runSetups().then(() => {
-  injectSystemCAs();
+app.on("ready", async () => {
```
Comment (jakolehm) [inline src/main/index.ts:65]: App start order was previously wrong... any async setups caused `app.on("ready")` registration happen too late -> we missed it completely.

File: src/main/proxy-functions/kube-api-upgrade-request.ts [hunk 17]
```diff
@@ -4,21 +4,25 @@
...
-import net from "net";
+import tls from "tls";
 import url from "url";
 import { apiKubePrefix } from "../../common/vars";
 import type { ProxyApiRequestArgs } from "./types";
 
 const skipRawHeaders = new Set(["Host", "Authorization"]);
 
-export async function kubeApiRequest({ req, socket, head, cluster }: ProxyApiRequestArgs) {
+export async function kubeApiUpgradeRequest({ req, socket, head, cluster }: ProxyApiRequestArgs) {
```
Comment (jakolehm) [inline src/main/proxy-functions/kube-api-upgrade-request.ts:14]: Renamed because name was misleading, only upgrade requests are handled here.

File: src/main/context-handler/context-handler.ts [hunk 18]
```diff
@@ -12,6 +12,9 @@ import url, { UrlWithStringQuery } from "url";
 import { CoreV1Api } from "@kubernetes/client-node";
 import logger from "../logger";
 import type { KubeAuthProxy } from "../kube-auth-proxy/kube-auth-proxy";
+import path from "path";
+import { readFile } from "fs/promises";
```
Comment (Nokel81) [inline src/main/context-handler/context-handler.ts]: This should probably be a dependency

File: src/main/context-handler/context-handler.ts [hunk 19]
```diff
@@ -118,7 +122,11 @@ export class ContextHandler {
     await this.ensureServer();
     const path = this.clusterUrl.path !== "/" ? this.clusterUrl.path : "";
 
-    return `http://127.0.0.1:${this.kubeAuthProxy.port}${this.kubeAuthProxy.apiPrefix}${path}`;
+    return `https://127.0.0.1:${this.kubeAuthProxy.port}${this.kubeAuthProxy.apiPrefix}${path}`;
+  }
+
+  async resolveAuthProxyCa() {
+    return readFile(path.join(this.dependencies.certPath, "proxy.crt"));
```
Comment (Nokel81) [inline src/main/context-handler/context-handler.ts]: If this file never changes I think it should be a setup-able dependency.

File: src/main/index.ts (hunk 16 above)
Comment (Nokel81) [reply src/main/index.ts:65]: So that we can catch these async errors I would suggest a slightly different shape:

```
async function main() {
  const di = getDi();

  await app.whenReady();
  await di.runSetups();

  ...
}

main().catch(error => {
  console.error(error);
  dialog.showErrorBox("Lens Startup Failed", String(error));
  app.exit(1);
});
```

File: src/main/kube-auth-proxy/create-kube-auth-proxy-cert-files.injectable.ts [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+import { getKubeAuthProxyCertificatePath } from "./kube-auth-proxy-cert";
+import * as selfsigned from "selfsigned";
+import { createKubeAuthProxyCertFiles } from "./create-kube-auth-proxy-cert-files";
+
+const createKubeAuthProxyCertFilesInjectable = getInjectable({
+  id: "create-kube-auth-proxy-cert-files",
+
+  instantiate: async (di) => {
+    const userData = di.inject(directoryForUserDataInjectable);
+    const certPath = getKubeAuthProxyCertificatePath(userData);
```
Comment (Nokel81) [inline src/main/kube-auth-proxy/create-kube-auth-proxy-cert-files.injectable.ts]: This sounds more like a dependency which should depend on `directoryForUserDataInjectable`.

File: src/main/kube-auth-proxy/create-kube-auth-proxy-cert-files.injectable.ts [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    
+    return createKubeAuthProxyCertFiles(certPath, {
```
Comment (Nokel81) [inline src/main/kube-auth-proxy/create-kube-auth-proxy-cert-files.injectable.ts:19]: Given that this creates the files and then returns `certPath` (which coincendentally should probably be called something like `directoryForProxyCertificates`. I think that this injectable should actually be mostly just `setup` and then instantiate should return `undefined` or maybe a new `SetupOnly` marker type.

Then whatever depends on `createKubeAuthProxyCertFilesInjectable` could instead just depend on `directoryForProxyCertificatesInjectable` and know that the setup and creation of the files will already have been done.

File: src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts [hunk 22]
```diff
@@ -4,18 +4,24 @@
...
 const createKubeAuthProxyInjectable = getInjectable({
   id: "create-kube-auth-proxy",
 
-  instantiate: (di) => {
+  instantiate: (di): CreateKubeAuthProxy => {
     const binaryName = getBinaryName("lens-k8s-proxy");
     const dependencies: KubeAuthProxyDependencies = {
       proxyBinPath: path.join(di.inject(directoryForBundledBinariesInjectable), binaryName),
+      proxyCertPath: di.inject(getKubeAuthProxyCertDirInjectable),
+      spawn,
```
Comment (Nokel81) [inline src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts]: This should be its own injectable, maybe within a folder `src/main/child-process` since we probably shouldn't have any of those calls within `renderer`

File: types/selfsigned.d.ts [hunk 23]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
+ declare module "selfsigned" {
```
Comment (Nokel81) [inline types/selfsigned.d.ts:6]: I think it would be better if these were hosted into `@types`. The PR process is actually really simple, I have done it a couple of times.

File: src/main/context-handler/context-handler.ts (hunk 19 above)
Comment (jakolehm) [reply src/main/context-handler/context-handler.ts]: Refactored. Now `authProxyCa` is a dependency (but still a promise because of `createContextHandlerInjectable` instantiate has to be non-async for now). This way `ContextHandler` does not need to know `readFile` at all.

File: src/main/context-handler/context-handler.ts (hunk 9 above)
Comment (jakolehm) [reply src/main/context-handler/context-handler.ts:141]: It is out of scope, topic for an another PR.

File: src/main/kube-auth-proxy/create-kube-auth-proxy-cert-files.injectable.ts (hunk 20 above)
Comment (jakolehm) [reply src/main/kube-auth-proxy/create-kube-auth-proxy-cert-files.injectable.ts]: Indeed, there was already an injectable for this. Refactored.

File: src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts (hunk 22 above)
Comment (jakolehm) [reply src/main/kube-auth-proxy/create-kube-auth-proxy.injectable.ts]: Done.

File: types/selfsigned.d.ts (hunk 23 above)
Comment (jakolehm) [reply types/selfsigned.d.ts:6]: It would be better yes but I think it can be separately from this PR. Note: this declaration is probably not complete, it only touches parts this PR needs.

**Comment by jakolehm** [comment]
@Nokel81 PTAL again.

File: src/main/lens-proxy.ts [hunk 24]
```diff
@@ -61,7 +61,7 @@ export class LensProxy extends Singleton {
 
   public port: number;
 
-  constructor(protected router: Router, protected proxy: httpProxy, { shellApiRequest, kubeApiRequest, getClusterForRequest }: LensProxyFunctions) {
+  constructor(protected router: Router, protected proxy: httpProxy, { shellApiRequest, kubeApiUpgradeRequest, getClusterForRequest }: LensProxyFunctions) {
```
Comment (Nokel81) [inline src/main/lens-proxy.ts:64]: nit: all of these feel like dependencies. Should we consolidate them?

File: src/renderer/components/delete-cluster-dialog/__tests__/delete-cluster-dialog.test.tsx [hunk 25]
```diff
@@ -91,6 +92,8 @@ describe("<DeleteClusterDialog />", () => {
   beforeEach(async () => {
     const { mainDi, runSetups } = getDisForUnitTesting({ doGeneralOverrides: true });
 
+    mainDi.override(createContextHandlerInjectable, () => () => undefined);
```
Comment (Nokel81) [inline src/renderer/components/delete-cluster-dialog/__tests__/delete-cluster-dialog.test.tsx:95]: Should this be done in `getDiForUnitTesting` so that it is always run by default?

File: src/main/lens-proxy.ts (hunk 24 above)
Comment (jakolehm) [reply src/main/lens-proxy.ts:64]: Probably yes but sounds like out of scope for this PR?

File: src/renderer/components/delete-cluster-dialog/__tests__/delete-cluster-dialog.test.tsx (hunk 25 above)
Comment (jakolehm) [reply src/renderer/components/delete-cluster-dialog/__tests__/delete-cluster-dialog.test.tsx:95]: It cannot be there because some tests do require the real implementation.

File: src/main/lens-proxy.ts (hunk 24 above)
Comment (Nokel81) [reply src/main/lens-proxy.ts:64]: Sure

File: src/renderer/components/delete-cluster-dialog/__tests__/delete-cluster-dialog.test.tsx (hunk 25 above)
Comment (Nokel81) [reply src/renderer/components/delete-cluster-dialog/__tests__/delete-cluster-dialog.test.tsx:95]: Oh yes, that is why I would put it behind some sort of flag (like `doGeneralOverrides` is). I suggested moving it because I think that the vast majority of tests would want it overridden.

**Review by Nokel81 — APPROVED** [review]
We can change how certain injectables are overridden by default in a future PR

**Review by nevalla — APPROVED** [review]
